analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs

```diff
@@ -0,0 +1,241 @@
+        }
+    }
+
+    private sealed class PropertyTypeMismatch
+    {
+        public TypeSyntax Type { get; }
+        public string ConstraintType { get; }
+        public string Route { get; }
+        public Location RouteParamLocation { get; }
+
+        private bool IsImplicitStringConstraint => ConstraintType.Equals("string", StringComparison.OrdinalIgnoreCase);
+        private bool CanReportSecondaryLocation => RouteParamLocation is not null;
+
+        public PropertyTypeMismatch(TypeSyntax type, string constraintType, string route, Location routeParamLocation)
+        {
+            Type = type;
+            ConstraintType = constraintType;
+            Route = route;
+            RouteParamLocation = routeParamLocation;
+        }
+
+        public string ToPrimaryMessage()
+        {
+            if (CanReportSecondaryLocation)
+            {
+                return string.Format(MessageWithSecondaryLocationFormat, GetTypeName(Type));
+            }
+
+            if (IsImplicitStringConstraint)
+            {
+                return string.Format(ImplicitStringPrimaryMessageFormat, GetTypeName(Type), Route);
+            }
+
+            return string.Format(PrimaryMessageFormat, GetTypeName(Type), ConstraintType.ToLower(), Route);
+        }
+
+        private static string GetTypeName(TypeSyntax type) =>
+            type switch
+            {
+                ArrayTypeSyntax _ => type.ToString(),
+                PointerTypeSyntax _ => type.ToString(),
+                _ => type.GetName()
+            };
+
+        public IEnumerable<Location> ToAdditionalLocation() =>
+            CanReportSecondaryLocation ? new[] { RouteParamLocation } : Array.Empty<Location>();
+
+        public ImmutableDictionary<string, string> ToProperties() => CanReportSecondaryLocation
+            ? new Dictionary<string, string> { { "0", ToSecondaryMessage() } }
+                .ToImmutableDictionary()
+            : ImmutableDictionary<string, string>.Empty;
+
+        private string ToSecondaryMessage() =>
+            IsImplicitStringConstraint ? ImplicitStringSecondaryMessage : string.Format(SecondaryMessageFormat, ConstraintType.ToLower());
+    }
+
+    private sealed record RouteParameter(string Constraint, string FromRoute, Location RouteParamLocation);
+
+    private sealed record SupportedType(KnownType ConstraintKnownType)
+    {
+        public bool Matches(ITypeSymbol propertyType, Compilation compilation)
+        {
+            if (propertyType.Is(ConstraintKnownType))
+            {
+                return true;
+            }
+
+            var constraintTypeSymbol = compilation.GetTypeByMetadataName(ConstraintKnownType);
+
+            var conversion = compilation.ClassifyConversion(constraintTypeSymbol, propertyType);
+
+            return conversion.IsBoxing
+                   || conversion.IsEnumeration;
+        }
+    }
```

Comment:
I have multiple route that use the same property `EnumParam`:

```razor
@page "/route/4/{EnumParam:int}" @* Compliant *@
@page "/route/5/{EnumParam:long}" @* Compliant *@
@page "/route/6/{EnumParam:guid}" @* Noncompliant *@
```

Since it is confusing, I will make it more explicit.